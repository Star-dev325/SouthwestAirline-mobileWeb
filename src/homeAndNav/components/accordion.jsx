// @flow
import React, { Component } from 'react';
import type { Element } from 'react';
import Icon from 'src/shared/components/icon';
import cx from 'classnames';

type Props = {
  open: boolean,
  icon: boolean,
  heading: Element<*>,
  body: Element<*>,
  onHeaderClick: () => void,
  className?: string
};

type State = {
  open: boolean,
  height: number | string,
  expandHeight: number | string,
  unexpandHeight: number
};

class Accordion extends Component<Props, State> {
  state = {
    open: this.props.open,
    height: 'auto',
    expandHeight: 'auto',
    unexpandHeight: 0
  };

  componentDidMount() {
    const { open } = this.state,
      // $FlowFixMe children type does not have a property `length`
      expandHeight = this.props.body.props.children.length * 38;

    if (open) {
      this.setState({
        expandHeight,
        height: expandHeight
      });
    } else {
      this.setState({
        expandHeight,
        height: this.state.unexpandHeight
      });
    }
  }

  UNSAFE_componentWillReceiveProps(nextProps: Props) {
    // $FlowFixMe children type does not have a property `length`
    const newHeight = nextProps.body.props.children.length * 38;

    this.setState({
      open: nextProps.open,
      expandHeight: newHeight,
      height: nextProps.open ? newHeight : this.state.unexpandHeight
    });
  }

  _renderHeading = (props: Props) => (
    <div onClick={this.handleHeaderClick} className="accordion--heading">
      {props.heading}

      {props.icon ? <Icon type={this.state.open ? 'openeddrawer' : 'closeddrawer'} /> : null}
    </div>
  );

  _renderBody = (props: Props) => {
    const { open, height } = this.state;
    const styleHeight = open ? 'auto' : height;

    return (
      <div style={{ height: styleHeight }} className="accordion--body" ref="accordionBody">
        {props.body}
      </div>
    );
  };

  handleHeaderClick = () => {
    const { unexpandHeight, expandHeight } = this.state,
      isOpen = !this.state.open,
      { onHeaderClick } = this.props;

    this.setState(
      {
        open: isOpen,
        height: isOpen ? expandHeight : unexpandHeight
      },
      onHeaderClick
    );
  };

  render() {
    const { props } = this;

    return (
      <div ref="container" className={cx('accordion', props.className)}>
        {this._renderHeading(props)}
        {this._renderBody(props)}
      </div>
    );
  }
}

export default Accordion;
