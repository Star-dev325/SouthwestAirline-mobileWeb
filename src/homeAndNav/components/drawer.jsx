// @flow
import React, { Component } from 'react';
import { connect } from 'react-redux';
import _ from 'lodash';
import cx from 'classnames';
import { toggleDrawer } from 'src/homeAndNav/actions/drawerActions';
import { getAppContent, addClass, removeClass } from 'src/shared/helpers/domUtils';

type Props = {
  children: *,
  appSizeGetter: () => {
    globalHeaderHeight: number,
    appScrollTop: number
  },
  toggleDrawerFn: (state: boolean) => void,
  isDrawerOpen: boolean,
  scrollDrawerToTop: boolean,
  isJourneyBannerDisplayed: boolean,
  appOffsetTop: number
};

type State = {
  open: boolean,
  containerHeight: string,
  top: number
};

export class Drawer extends Component<Props, State> {
  state: State = {
    open: this.props.isDrawerOpen,
    containerHeight: '0',
    top: 0
  };

  componentDidMount() {
    this._toggleScrollOnAppContent();
    this._onStatusChange(this.props.scrollDrawerToTop);
  }

  componentDidUpdate(prevProps: Props) {
    this._toggleScrollOnAppContent();

    if (this.props.isJourneyBannerDisplayed !== prevProps.isJourneyBannerDisplayed) {
      this._setDrawerSizeByAppSize();
    }
  }

  _toggleScrollOnAppContent = () => {
    const appContent = getAppContent();
    const disableScrollingClass = 'disable-scrolling';

    if (!appContent) {
      return;
    }

    if (this.props.isDrawerOpen) {
      addClass(appContent, disableScrollingClass);
      this._onStatusChange(this.props.scrollDrawerToTop);
    } else {
      removeClass(appContent, disableScrollingClass);
    }
  };

  _setDrawerSizeByAppSize = () => {
    const { globalHeaderHeight = 0, appScrollTop = 0 } = this.props.appSizeGetter();
    const { appOffsetTop } = this.props;
    const drawerTop = appScrollTop > globalHeaderHeight ? 0 : globalHeaderHeight - appScrollTop + appOffsetTop;

    this.setState({
      containerHeight: `calc(100% - ${drawerTop}px)`,
      top: drawerTop
    });
  };

  _onStatusChange = (scrollTop: boolean) => {
    const { open } = this.state;

    if (open !== this.props.isDrawerOpen) {
      if (this.props.isDrawerOpen) {
        this._setDrawerSizeByAppSize();
        this.resetScrollTop();
      }
      this.setState({
        open: this.props.isDrawerOpen
      });
    }
    !scrollTop && this.resetScrollTop();
  };

  resetScrollTop = () => {
    this.refs.drawerContent.scrollTop = 0;
  };

  closeDrawer = () => {
    const { isDrawerOpen, toggleDrawerFn } = this.props;

    if (isDrawerOpen) {
      toggleDrawerFn(true);
    }
  };

  render() {
    const active = this.props.isDrawerOpen ? 'active' : '';
    const { top, containerHeight } = this.state;
    const debouncedCloseDrawerOnTouchMove = _.debounce((...args) => this.closeDrawer(...args), 1000, {
      leading: true,
      trailing: false
    });

    return (
      <div className={cx(active, 'drawer')} ref="drawer" style={{ top }}>
        <div
          className="drawer--bg disable-scrolling"
          onClick={() => this.closeDrawer()}
          ref="drawerBg"
          onTouchMove={debouncedCloseDrawerOnTouchMove}
        />
        <div id="drawerContent" className="drawer--content" ref="drawerContent" style={{ height: containerHeight }}>
          {this.props.children}
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  isDrawerOpen: state.app.homeAndNav.drawer.isDrawerOpen,
  scrollDrawerToTop: state.app.homeAndNav.drawer.scrollDrawerToTop
});

const mapDispatchToProps = {
  toggleDrawerFn: toggleDrawer
};

export default connect(mapStateToProps, mapDispatchToProps)(Drawer);
