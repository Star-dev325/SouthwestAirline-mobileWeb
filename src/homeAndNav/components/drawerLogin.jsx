// @flow

import React from 'react';
import _ from 'lodash';
import cx from 'classnames';

import TierTypes from 'src/shared/constants/tierTypes';
import UserPreferName from 'src/homeAndNav/components/userPreferName';
import Icon from 'src/shared/components/icon';
import i18n from '@swa-ui/locale';
import { formatCurrency } from 'src/shared/helpers/formatCurrencyHelper';

import type { UserInfo } from 'src/homeAndNav/flow-typed/homeAndNav.types';

type Props = {
  userInfo?: UserInfo,
  isLoggedIn: boolean,
  onDrawerLoginClicked: (to: string, params: ?Array<mixed>, query: ?{ to: string }) => void,
  toggleDrawer: (state: boolean) => void
};

const DrawerLogin = (props: Props) => {
  const { userInfo, isLoggedIn } = props;

  const getAccountInfo = () => {
    if (isLoggedIn && userInfo && !!userInfo.accountInfo) {
      const { accountInfo } = userInfo;

      return {
        name: accountInfo.customerInfo.name.userName,
        pts: formatCurrency(accountInfo.rapidRewardsDetails.redeemablePoints, false),
        tier: accountInfo.rapidRewardsDetails.tierInfo.tier,
        isTierStatusPending: accountInfo.isTierStatusPending,
        isEnrolledInRapidRewards: accountInfo.rapidRewardsDetails.isEnrolledInRapidRewards
      };
    }
  };

  const renderAccountInfo = () => {
    const accountInfo = getAccountInfo();
    let content = null;

    if (accountInfo) {
      const showRapidRewardsPoints = _shouldShowRapidRewardsPoints();
      const shouldShowTierLevel = _shouldShowTierLevel();

      content = (
        <div>
          <div className="flex flex-main-between xxlarge bold">
            <div data-qa="drawer-login-customer-name" className="overflow-hidden nowrap ellipsis">
              <UserPreferName isLoggedIn={isLoggedIn} name={accountInfo.name} />
            </div>
            {showRapidRewardsPoints && <div className="align-right bold flex-shrink0">{accountInfo.pts}</div>}
          </div>
          <div className="flex flex-main-between larger">
            {shouldShowTierLevel && <p>{TierTypes[accountInfo.tier]}</p>}
            {showRapidRewardsPoints && (
              <p className={cx('align-right', { 'flex-auto': !shouldShowTierLevel })}>{i18n('HOME_AND_NAV__PTS')}</p>
            )}
          </div>
          <div data-qa="nav-drawer-view-account" className="flex larger flex-main-end flex-cross-center pt3 pblue">
            {i18n('HOME_AND_NAV__NAV_DRAWER__VIEW_ACCOUNT')}
            <Icon type="keyboard-arrow-right" className="xxlarge" />
          </div>
        </div>
      );
    }

    return content;
  };

  const _shouldShowRapidRewardsPoints = () => {
    const accountInfo = getAccountInfo();
    const enrolledInRapidRewards = _.get(accountInfo, 'isEnrolledInRapidRewards', false);

    return enrolledInRapidRewards;
  };

  const _shouldShowTierLevel = () => {
    const accountInfo = getAccountInfo();
    const enrolledInRapidRewards = _.get(accountInfo, 'isEnrolledInRapidRewards', false);
    const tierStatusPending = _.get(accountInfo, 'isTierStatusPending', false);

    return enrolledInRapidRewards && !tierStatusPending;
  };

  const _onClicked = () => {
    const { onDrawerLoginClicked, toggleDrawer } = props;
    const params = isLoggedIn ? ['/my-account?clk=GNAVMYACCT2'] : ['/login?clk=GNAVLOGIN2', null, { to: '/my-account' }]; 

    onDrawerLoginClicked(...params);
    toggleDrawer(true);
  };

  return (
    <div className="drawer-login" onClick={_onClicked}>
      <div className={cx('drawer-login--placeholder', { hide: isLoggedIn })}>
        <Icon type="profile" className="drawer-login--icon_profile" />
        <div className="drawer-login--reminder">
          <div>
            <div className="bold inline">{i18n('HOME_AND_NAV__NAV_DRAWER__LOGIN_OR_ENROLL')}</div>
            <div className="inline">{i18n('HOME_AND_NAV__NAV_DRAWER__TO_MANAGE')}</div>
          </div>
          <div>{i18n('HOME_AND_NAV__NAV_DRAWER__YOUR_TRIPS')}</div>
        </div>
      </div>
      {renderAccountInfo()}
    </div>
  );
};

export default DrawerLogin;
