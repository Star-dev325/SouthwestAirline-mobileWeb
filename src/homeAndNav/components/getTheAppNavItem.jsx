// @flow

import React from 'react';
import RichNavItem from 'src/shared/components/richNavItem';
import DeviceInfo from 'src/shared/helpers/deviceInfo';
import NavDrawerConstants from 'src/homeAndNav/constants/navDrawerConstants';

const { ANDROID_URL, IOS_URL } = NavDrawerConstants;

type Props = {
  onClick: ({ link_type: string, target: string }) => void
};

const GetTheAppNavItem = (props: Props) => {
  const { onClick } = props;
  const shouldNotRender = DeviceInfo.os.name !== 'Android' && DeviceInfo.os.name !== 'iOS';

  if (shouldNotRender) {
    return null;
  }

  const getTheAppImageRender = () => <div className="home-nav-list-item--image image-promo--swa-app" />;
  const getTheAppItem = {
    title: 'Get the app',
    description: 'Retrieve reservation, checkin for upcoming flights & more',
    link_type: 'browser',
    target: DeviceInfo.os.name === 'Android' ? ANDROID_URL : IOS_URL
  };

  return (
    <RichNavItem
      className="home-nav-list-item"
      imageRender={getTheAppImageRender}
      onClick={onClick}
      {...getTheAppItem}
    />
  );
};

export default GetTheAppNavItem;
