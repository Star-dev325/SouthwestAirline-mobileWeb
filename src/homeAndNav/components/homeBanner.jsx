// @flow
import React from 'react';
import { isEmpty } from 'src/shared/helpers/jsUtils';
import DynamicPlacement from 'src/wcm/components/dynamicPlacement';
import { HOME_PAGE_ID } from 'src/wcm/constants/wcmConstants';

type Props = {
  handleFirmOfferOfCreditFn: () => void,
  homeBanners: Array<*>
};

const HomeBanner = ({ homeBanners, handleFirmOfferOfCreditFn }: Props) => {
  if (isEmpty(homeBanners)) {
    return null;
  }

  return (
    <div className="home-banner">
      {homeBanners.map((content, index) => (
        <DynamicPlacement
          data-qa="home-banner-item"
          key={index}
          observerCallback={handleFirmOfferOfCreditFn}
          pageId={HOME_PAGE_ID}
          {...content}
        />
      ))}
    </div>
  );
};

export default HomeBanner;
