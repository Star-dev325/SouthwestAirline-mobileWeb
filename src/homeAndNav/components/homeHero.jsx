// @flow
import React from 'react';
import type { Node } from 'react';
import _ from 'lodash';
import DynamicPlacement from 'src/wcm/components/dynamicPlacement';
import { HOME_PAGE_ID } from 'src/wcm/constants/wcmConstants';

type Props = {
  heroContents: Array<*>,
  handleFirmOfferOfCreditFn: () => void
};

const HomeHero = ({ heroContents, handleFirmOfferOfCreditFn }: Props): Node => {
  if (_.isEmpty(heroContents)) {
    return null;
  }

  return heroContents.map((content, index) => (
    <DynamicPlacement
      data-qa="home-hero-item"
      key={index}
      observerCallback={handleFirmOfferOfCreditFn}
      pageId={HOME_PAGE_ID}
      {...content} />
  ));
};

export default HomeHero;
