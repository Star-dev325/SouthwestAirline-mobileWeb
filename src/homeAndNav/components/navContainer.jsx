// @flow

import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import MenuList from 'src/homeAndNav/components/menuList';
import DrawerLogin from 'src/homeAndNav/components/drawerLogin';
import MenuFooter from 'src/homeAndNav/components/menuFooter';
import { toggleDrawer } from 'src/homeAndNav/actions/drawerActions';
import { updateContentBlockIdsFromMenuList } from 'src/shared/analytics/actions/analyticsActions';
import { raiseSatelliteEvent } from 'src/shared/analytics/helpers/analyticsEventHelper';

import type { NavClickItem, UserInfo } from 'src/homeAndNav/flow-typed/homeAndNav.types';

type Props = {
  userInfo?: UserInfo,
  onNavClick: (NavClickItem) => void,
  isLoggedIn: boolean,
  onLoginClick: (NavClickItem) => void,
  onLogoutClick: () => void,
  toggleDrawerFn: (boolean) => void,
  menuList: Array<*>,
  isDrawerOpen: boolean,
  updateContentBlockIdsFromMenuListFn: (Array<*>) => void
};

export const NavContainer = ({
  userInfo,
  isLoggedIn,
  onLoginClick,
  onLogoutClick,
  onNavClick,
  toggleDrawerFn,
  menuList,
  isDrawerOpen,
  updateContentBlockIdsFromMenuListFn
}: Props) => {
  useEffect(() => {
    if (isDrawerOpen) {
      updateContentBlockIdsFromMenuListFn(menuList);
      raiseSatelliteEvent('overlay:hamburger navigation');
    }
  }, [isDrawerOpen]);

  const _onDrawerLoginClicked = (to, params, query) => {
    const navItem = {
      routeName: to,
      params,
      query,
      isWcmLink: false
    };

    onLoginClick(navItem);
  };

  const _onLinkClick = (item) => {
    onNavClick(item);
    toggleDrawerFn(true);
  };

  return (
    <div className="nav-container">
      <DrawerLogin
        userInfo={userInfo}
        isLoggedIn={isLoggedIn}
        onDrawerLoginClicked={_onDrawerLoginClicked}
        toggleDrawer={toggleDrawerFn}
      />
      <MenuList
        onLinkClick={_onLinkClick}
        isLoggedIn={isLoggedIn}
        listData={menuList}
        onLogoutClick={onLogoutClick}
        toggleDrawer={toggleDrawerFn}
      />
      <MenuFooter />
    </div>
  );
};

const mapStateToProps = (state) => ({
  isLoggedIn: state.app.account.isLoggedIn,
  userInfo: state.app.account,
  menuList: state.app.homeAndNav.menuList.listData,
  isDrawerOpen: state.app.homeAndNav.drawer.isDrawerOpen
});
const mapDispatchToProps = {
  toggleDrawerFn: toggleDrawer,
  updateContentBlockIdsFromMenuListFn: updateContentBlockIdsFromMenuList
};

export default connect(mapStateToProps, mapDispatchToProps)(NavContainer);
