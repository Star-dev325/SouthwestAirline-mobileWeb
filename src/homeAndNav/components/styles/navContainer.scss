// OK to declare here I think because this is only place they're used
// move to _colors.scss if they ever need to be re-used
$primary-facebook: #3b5998;
$primary-twitter: #00acee;

.nav-container {
  @include bgwhite;
  @include larger;

  li {
    list-style-type: none;
  }

  a {
    @include pdkblue;
  }

  .icon_call-us,
  .icon_email-us,
  .icon_ic-feedback,
  .icon_twitter-2,
  .icon_facebook-2,
  .icon_chat-with-us {
    @include pl8;
    @include pr4;
    font-size: 3.6rem;
  }

  .icon_call-us {
    @include pblue;
  }

  .icon_email-us {
    @include sdkblue;
  }

  .icon_twitter-2 {
    color: $primary-twitter;
  }

  .icon_facebook-2 {
    color: $primary-facebook;
  }

  .icon_ic-feedback {
    @include sblue;
  }

  .menu-list {
    padding-left: 1.5rem;
    padding-right: 1.5rem;

    .dynamic-placement {
      margin: 0 -1.5rem;
    }
  }

  .menu-footer {
    border-top: 1px solid $gray2;
    padding: 2rem 1.5rem 1rem;

    .menu-footer--copyright {
      @include block;
      @include gray5;
      @include lineheight14;
      @include center;
      font-size: 1.3rem;
    }
  }
}
