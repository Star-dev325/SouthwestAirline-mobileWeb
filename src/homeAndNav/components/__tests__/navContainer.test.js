jest.mock('src/shared/helpers/deviceInfo', () => ({
  os: {
    name: 'iOS'
  }
}));
jest.mock('src/shared/config/appConfig', () => ({
  userCanChangeToggles: jest.fn().mockReturnValue(true)
}));
import { fireEvent } from '@testing-library/react';
import configureMockStore from 'redux-mock-store';
import * as DrawerActions from 'src/homeAndNav/actions/drawerActions';
import { NavContainer } from 'src/homeAndNav/components/navContainer';
import * as MenuListData from 'src/homeAndNav/constants/menuListData';
import * as analyticsEventHelper from 'src/shared/analytics/helpers/analyticsEventHelper';
import deviceInfo from 'src/shared/helpers/deviceInfo';
import WcmMenusTransformedBuilder from 'test/builders/model/wcmMenusTransformedBuilder';
import { integrationRender } from 'test/unit/helpers/testingLibraryUtils';

describe('NavContainer', () => {
  let listData;
  let onLoginClickMock;
  let onNavClickMock;
  let toggleDrawerFnMock;
  let updateContentBlockIdsFromMenuListFnMock;

  beforeEach(() => {
    jest.spyOn(analyticsEventHelper, 'raiseSatelliteEvent');
    onLoginClickMock = jest.fn();
    onNavClickMock = jest.fn();
    toggleDrawerFnMock = jest.spyOn(DrawerActions, 'toggleDrawer');
    updateContentBlockIdsFromMenuListFnMock = jest.fn();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  function renderNavContainer(props = {}, menuData) {
    const listData = menuData ? menuData : MenuListData.MenuListData;
    const defaultProps = {
      isDrawerOpen: false,
      isLoggedIn: false,
      isLogin: false,
      menuList: listData,
      onLoginClick: onLoginClickMock,
      onLogoutClick: () => {},
      onNavClick: onNavClickMock,
      toggleDrawerFn: toggleDrawerFnMock,
      updateContentBlockIdsFromMenuListFn: updateContentBlockIdsFromMenuListFnMock,
      userInfo: {
        isLogin: false,
        name: 'test'
      }
    };

    const componentProps = { ...defaultProps, ...props };

    const state = configureMockStore()({
      app: {
        account: {
          isLoggedIn: false
        },
        errorHeader: {
          errorMessage: null,
          hasError: false
        },
        homeAndNav: {
          menuList: {
            activeMenuIndex: 1,
            listData: componentProps.menuList
          }
        }
      }
    });

    return integrationRender()(state, NavContainer, componentProps);
  }

  beforeEach(() => {
    listData = getListData();
  });

  afterEach(() => {
    jest.clearAllMocks();
    deviceInfo.os.name = 'iOS';
  });

  describe('contact label', () => {
    describe('multiple mounting', () => {
      it('should show only one get the app link', () => {
        deviceInfo.os.name = 'Android';
        listData = getListData('Android');

        // Multiple mount of this page
        const { unmount } = renderNavContainer({}, listData);

        unmount();

        const { container } = renderNavContainer({}, listData);

        expect(container.querySelectorAll('[data-qa="get-the-mobile-app"]')).toHaveLength(1);
      });
    });

    describe('android', () => {
      it('should have get the mobile app label pointing to the play store', () => {
        deviceInfo.os.name = 'Android';
        listData = getListData('Android');

        const { container } = renderNavContainer({}, listData);
        const pageComponent = container.querySelector('[data-qa="get-the-mobile-app"]');

        fireEvent.click(pageComponent.querySelector('.menu-list-item--heading'));

        expect(onNavClickMock.mock.calls[0][0].link).toEqual(
          'https://play.google.com/store/apps/details?id=com.southwestairlines.mobile&hl=en'
        );
        expect(toggleDrawerFnMock).toHaveBeenCalledWith(true);
      });
    });

    describe('ios', () => {
      it('should have get the ios app label pointing to the app store', () => {
        listData = getListData('iOS');

        const { container } = renderNavContainer({}, listData);
        const pageComponent = container.querySelector('[data-qa="get-the-mobile-app"]');

        fireEvent.click(pageComponent.querySelector('.menu-list-item--heading'));

        expect(onNavClickMock.mock.calls[0][0].link).toEqual(
          'https://itunes.apple.com/us/app/southwest-airlines/id344542975'
        );
        expect(toggleDrawerFnMock).toHaveBeenCalledWith(true);
      });
    });

    describe('other OS', () => {
      it('should not have get the ios app label', () => {
        deviceInfo.os.name = 'Windows';
        listData = getListData('Windows');

        const { container } = renderNavContainer({}, listData);

        expect(container.querySelector('[data-qa="get-the-mobile-app"]')).toBeNull();
      });
    });
  });

  describe('#onLinkClick', () => {
    it('should pass onLogoutClick function to MenuList', () => {
      const { container } = renderNavContainer({}, listData);

      expect(container).toMatchSnapshot();
    });

    it('should pass query parameters to transition function', () => {
      const { container } = renderNavContainer({}, listData);

      fireEvent.click(container.querySelector('[data-link="/view-reservation"]'));

      expect(onNavClickMock.mock.calls[0][0].routeName).toEqual('/view-reservation');
      expect(onNavClickMock.mock.calls[0][0].query).toEqual({ clk: 'GNAVCARVIEWRES', tab: 'CAR' });
    });
  });

  describe('#onDrawerLoginClicked', () => {
    it('should pass query parameters to onLoginClick', () => {
      const { container } = renderNavContainer({}, listData);

      fireEvent.click(container.querySelector('.drawer-login'));

      expect(onLoginClickMock).toHaveBeenCalled();
    });
  });

  describe('feature toggles', () => {
    describe('when in an environment where user can change them', () => {
      afterEach(() => {
        jest.clearAllMocks();
      });

      describe('the feature toggle page link', () => {
        let featureTogglePageLink;

        it('should exist', () => {
          const { container } = renderNavContainer({}, listData);

          featureTogglePageLink = container.querySelectorAll('.container')[1];

          expect(featureTogglePageLink.textContent).toContain('CHANGE TOGGLES');
        });

        it('should transition to the feature toggle page when clicked', () => {
          const { container } = renderNavContainer({}, listData);

          featureTogglePageLink = container.querySelectorAll('.container')[1];

          fireEvent.click(featureTogglePageLink.querySelector('.menu-list-item--heading'));

          expect(onNavClickMock.mock.calls[0][0].routeName).toEqual('/feature-toggles');
        });
      });

      describe('the environment settings page link', () => {
        let environmentSettingLink;

        it('should exist', () => {
          const { container } = renderNavContainer({}, listData);

          environmentSettingLink = container.querySelectorAll('.container')[2];

          expect(environmentSettingLink.textContent).toContain('ENVIRONMENT SETTINGS');
        });

        it('should transition to the view app config page when clicked', () => {
          const { container } = renderNavContainer({}, listData);

          environmentSettingLink = container.querySelectorAll('.container')[2];

          fireEvent.click(environmentSettingLink.querySelector('.menu-list-item--heading'));

          expect(onNavClickMock.mock.calls[0][0].routeName).toEqual('/view-app-config');
        });
      });
    });
  });

  describe('analytics', () => {
    it('should be called when drawer is open', () => {
      const props = {
        isDrawerOpen: true
      };

      renderNavContainer(props, listData);

      expect(updateContentBlockIdsFromMenuListFnMock).toHaveBeenCalledWith(listData);
      expect(analyticsEventHelper.raiseSatelliteEvent).toHaveBeenCalledWith('overlay:hamburger navigation');
    });
  });
});

const getListData = (deviceOS = 'undefined') =>
  new WcmMenusTransformedBuilder().withGetMobileAppMenu(deviceOS).withChangeTogglesMenu().withEnvSettingsMenu().build();
