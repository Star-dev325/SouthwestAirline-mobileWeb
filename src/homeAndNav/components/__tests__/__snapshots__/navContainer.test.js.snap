// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavContainer #onLinkClick should pass onLogoutClick function to MenuList 1`] = `
<div>
  <div>
    <div>
      <div
        class="nav-container"
      >
        <div
          class="drawer-login"
        >
          <div
            class="drawer-login--placeholder"
          >
            <i
              class="drawer-login--icon_profile icon icon_profile"
            />
            <div
              class="drawer-login--reminder"
            >
              <div>
                <div
                  class="bold inline"
                >
                  HOME_AND_NAV__NAV_DRAWER__LOGIN_OR_ENROLL
                </div>
                <div
                  class="inline"
                >
                  HOME_AND_NAV__NAV_DRAWER__TO_MANAGE
                </div>
              </div>
              <div>
                HOME_AND_NAV__NAV_DRAWER__YOUR_TRIPS
              </div>
            </div>
          </div>
        </div>
        <div
          class="menu-list"
        >
          <div
            class="container menu-list--border-bottom menu-list-item--heading-title-nav-bold"
            data-qa="menu-list-home"
          >
            <div
              class="menu-list-item--heading float-none"
            >
              <i
                class="menu-list-item--heading-icon mr4 icon icon_home"
              />
              <span
                class="menu-list-item--heading-title"
              >
                Home
              </span>
            </div>
          </div>
          <div
            class="accordion menu-list--border-bottom menu-list-item--heading-title-nav-bold"
          >
            <div
              class="accordion--heading"
            >
              <div
                class="menu-list-item--heading"
              >
                <i
                  class="menu-list-item--heading-icon mr4 icon icon_car"
                />
                <span
                  class="menu-list-item--heading-title-nav-bold"
                >
                  Car
                </span>
              </div>
              <i
                class="icon icon_openeddrawer"
              />
            </div>
            <div
              class="accordion--body"
              style="height: auto;"
            >
              <ul
                class="menu-list-item--body"
              >
                <div
                  class="menu-list-item--body-item"
                >
                  <li
                    class="inline"
                    data-link="/car/booking"
                    data-qa="menu-list-car-book-a-car"
                  >
                    Book a Car
                  </li>
                </div>
                <div
                  class="menu-list-item--body-item"
                >
                  <li
                    class="inline"
                    data-link="/view-reservation"
                    data-qa="menu-list-car-lookup-reservations"
                  >
                    Look Up Reservations
                  </li>
                </div>
              </ul>
            </div>
          </div>
          <div
            class="container menu-list-item--heading-title-nav-normal-no-children"
          >
            <div
              class="menu-list-item--heading float-none"
            >
              <span
                class="menu-list-item--heading-title-nav-normal-no-children"
              >
                CHANGE TOGGLES
              </span>
            </div>
          </div>
          <div
            class="container menu-list-item--heading-title-nav-normal-no-children"
          >
            <div
              class="menu-list-item--heading float-none"
            >
              <span
                class="menu-list-item--heading-title-nav-normal-no-children"
              >
                ENVIRONMENT SETTINGS
              </span>
            </div>
          </div>
        </div>
        <div
          class="menu-footer"
        >
          <span
            class="menu-footer--copyright"
          >
            Â©
            2024
             
            HOME_AND_NAV__MENU_LIST__SOUTHWEST_AIRLINES
          </span>
          <span
            class="menu-footer--copyright"
          >
            HOME_AND_NAV__MENU_LIST__RIGHTS_RESERVED
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
